### PR TITLE
external_api: add start/stop recording commands

### PR DESCRIPTION
**Pull request by horymury**
@saghul @damencho

### CODE REVIEW

File: modules/API/API.js [hunk 1]
```diff
@@ -13,6 +13,9 @@ import {
 import { parseJWTFromURLParams } from '../../react/features/base/jwt';
 import { invite } from '../../react/features/invite';
 import { toggleTileView } from '../../react/features/video-layout';
+import { getActiveSession } from '../../react/features/recording/functions';
```
Comment (saghul) [inline modules/API/API.js]: style: alpha sort

File: modules/API/API.js [hunk 2]
```diff
@@ -166,6 +169,26 @@ function initCommands() {
             } catch (err) {
                 logger.error('Failed sending endpoint text message', err);
             }
+        },
+        'start-recording': () => {
+            const { conference } = APP.store.getState()['features/base/conference'];
+
+            conference && conference.startRecording({
+                mode: JitsiRecordingConstants.mode.FILE,
```
Comment (saghul) [inline modules/API/API.js]: why is this not configurable?

File: modules/API/API.js [hunk 3]
```diff
@@ -166,6 +169,26 @@ function initCommands() {
...
+                appData: JSON.stringify({
+                    'file_recording_metadata': {
+                        'share': false
```
Comment (saghul) [inline modules/API/API.js]: ditto

File: modules/API/API.js [hunk 4]
```diff
@@ -166,6 +169,26 @@ function initCommands() {
...
+                    }
+                })
+            });
+        },
+        'stop-recording': () => {
+            const { conference } = APP.store.getState()['features/base/conference'];
```
Comment (saghul) [inline modules/API/API.js]: get the state on an auxiliary var please, instead of fetching it twice

**Comment by luixxiul** [comment]
This needs rebase too.

**Comment by jlrpuma** [comment]
@saghul do you have any idea when this PR will be merged ? it's a very useful functionality, I was looking for something like this but actually there's nothing that let me record the videoconference from the external api.

**Comment by horymury** [comment]
> @saghul do you have any idea when this PR will be merged ? it's a very useful functionality, I was looking for something like this but actually there's nothing that let me record the videoconference from the external api.

hi @jlrpuma, I'm sorry I left this PR hanging for a while, only got back to it last Friday. I'l update my PR with the addressed changes most probably today or tomorrow.

File: modules/API/API.js (hunk 4 above)
Comment (horymury) [reply modules/API/API.js]: done

File: modules/API/API.js (hunk 3 above)
Comment (horymury) [reply modules/API/API.js]: done

File: modules/API/API.js (hunk 2 above)
Comment (horymury) [reply modules/API/API.js]: done

File: modules/API/API.js (hunk 1 above)
Comment (horymury) [reply modules/API/API.js:18]: hmm there is only one function imported

**Comment by saghul** [comment]
@horymury In case you haven't, can you please rebase?

**Comment by horymury** [comment]
> @horymury In case you haven't, can you please rebase?

@saghul Rebased, also fixed some eslint issues.

File: modules/API/API.js (hunk 1 above)
Comment (saghul) [reply modules/API/API.js:18]: we sort by import path

File: modules/API/API.js [hunk 5]
```diff
@@ -15,6 +15,12 @@ import { setE2EEKey } from '../../react/features/e2ee';
 import { invite } from '../../react/features/invite';
 import { toggleTileView } from '../../react/features/video-layout';
 import { setVideoQuality } from '../../react/features/video-quality';
+import { getActiveSession } from '../../react/features/recording/functions';
+import { RECORDING_TYPES } from '../../react/features/recording/constants';
+import { JitsiRecordingConstants } from '../../react/features/base/lib-jitsi-meet';
+import {
+    isEnabled as isDropboxEnabled
+} from '../../react/features/dropbox';
```
Comment (saghul) [inline modules/API/API.js:23]: this should fit in a single line

File: modules/API/API.js [hunk 6]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
             logger.debug('Set video quality command received');
             sendAnalytics(createApiEvent('set.video.quality'));
             APP.store.dispatch(setVideoQuality(frameHeight));
+        },
+
+        /**
+         * Starts a file recording or streaming depending on the passed on params.
+         * - for youtube streams, recording `mode` should be `stream` and `youtubeStreamKey` must be passed on.
```
Comment (saghul) [inline modules/API/API.js:200]: look at the warnings above

File: modules/API/API.js [hunk 7]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
...
+         * @param { string } arg.dropboxToken - dropbox oauth2 token.
+         * @param { boolean } arg.shouldShare - whether the recording should be shared with the participants or not.
+         * Only applies to certain jitsi meet deploys.
+         * @param { string } arg.youtubeStreamKey - the youtube stream key.
+         * @param { string } arg.youtubeBroadcastID - the youtube broacast ID.
+         * @returns {void}
+         */
+        'start-recording': ({ mode, dropboxToken, shouldShare, youtubeStreamKey, youtubeBroadcastID }) => {
+            const state = APP.store.getState();
+            const { conference } = state['features/base/conference'];
```
Comment (saghul) [inline modules/API/API.js:217]: use `getCurrentConference`

File: modules/API/API.js [hunk 8]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
...
+                    mode: JitsiRecordingConstants.mode.STREAM,
+                    streamId: youtubeStreamKey
+                };
+            } else {
+                logger.error('Invalid recording mode provided');
+
+                return;
+            }
+
+            conference && conference.startRecording(recordingConfig);
```
Comment (saghul) [inline modules/API/API.js:268]: check if there is a conferene the first thing

File: modules/API/API.js [hunk 9]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
...
+        },
+
+        /**
+         * Stops a recording or streaming in progress.
+         *
+         * @param {string} mode - `file` or `stream`
```
Comment (saghul) [inline modules/API/API.js:274]: End the sentene in a period. I think the warning above is because of it.

File: modules/API/API.js [hunk 10]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
...
+         */
+        'stop-recording': mode => {
+            if (![ JitsiRecordingConstants.mode.FILE, JitsiRecordingConstants.mode.STREAM ].includes(mode)) {
+                logger.error('Invalid recording mode provided!');
+
+                return;
+            }
+
+            const state = APP.store.getState();
+            const { conference } = state['features/base/conference'];
```
Comment (saghul) [inline modules/API/API.js:285]: Use `getCurrentConference`

File: modules/API/API.js [hunk 11]
```diff
@@ -187,6 +193,102 @@ function initCommands() {
...
+
+            const activeSession = getActiveSession(state, mode);
+
+            if (activeSession && activeSession.id && conference) {
```
Comment (saghul) [inline modules/API/API.js:289]: first check for the conference

File: modules/API/API.js (hunk 1 above)
Comment (horymury) [reply modules/API/API.js]: done

**Comment by horymury** [comment]
@saghul think this is ready.

**Comment by jlrpuma** [comment]
@horymury on this [(https://github.com/horymury/jitsi-meet/pull/1)] i solve the conflicts on this PR could you please merge the changes in order to finish with this feature, thanks.

**Comment by jlrpuma** [comment]
Hi @saghul could you please review this PR?

**Comment by saghul** [comment]
Sorry for the delay!

**Comment by jlrpuma** [comment]
Hi @saghul thanks, how can i get the name of the folder that is generated (i guess by jibri) when i call the event _startRecording_, i mean, the folder where is stored the video when you have a local instalation of Jitsi. On my server im getting a folder with a random name '_emmnitbwyehgxaak/_' and there is stored the video, but i need to know how/where is generated that folder name ?

**Pull request merged**